File: Data Structures & Algorithms/longest-increasing-subsequence/submission-2.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [1] * n
        for i in range(1,n):
            for j in range(i):
                if (nums[i] > nums[j]):
                    dp[i] = max(dp[i], dp[j]+1)
        return max(dp)
# 9, 1, 4, 2, 3, 4, 7
# 1, 1, 2, 2, 3, 3, 4

File: Data Structures & Algorithms/two-integer-sum/submission-2.java
class Solution {

// why this works: so when you're passing through the array,
// the map will be empty until you encounter that value
// you try to store the actual value, and then if you can get it 
// through the difference, you get the pair
// [3, 4, 5, 6] target = 8
// so map would be :  {} initially
// iter 1: {3:1}
// iter 2: {4:2}; 8 - 4 = 4
// iter 3: {5:3}; 8 - 5 = 3, so then 5 + 3 = 8

    public int[] twoSum(int[] nums, int target) {
        int n = nums.length;
        int[] idx = new int[2];
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int i = 0 ; i  < n; i++){
            if(map.containsKey(target - nums[i])){
                idx[0] = map.get(target-nums[i]);
                idx[1] = i;
                break;
            }
            map.put(nums[i], i);
        }
        return idx;
    }
}
